Bela/main.cpp

```diff
@@ -199,6 +203,71 @@ struct DigiIO : csnd::InPlug<3> {
   }
 };
 
+/*  TrillIn opcode 
+  read values from Trill sensors
+  array sizes must be same as inumTouches
+  kactiveTouches, ksizes[], kvertLocations[], khoriLocations[] trill inumTouches, itrillID
+*/
+struct TrillIn : csnd::Plugin<4,2> {
+  int activeTouches = 0;
+  int numTouches;
+  int trillID;
+  float *touchSize; 
+  float *touchVertLocation;
+  float *touchHoriLocation;
+  
+  int init() {
+    numTouches = inargs[0];
+    trillID = inargs[1];
+    if (gTouchSensors.size() < trillID+1)
+      return csound->init_error("No Trill sensor connected");
+
+    touchSize = new float[numTouches];
+    touchVertLocation = new float[numTouches];
+    touchHoriLocation = new float[numTouches];    
+    return OK;
+  }
+  
+  int kperf () {
+    activeTouches = gTouchSensors[trillID]->getNumTouches();
+    outargs[0] = MYFLT (activeTouches);
+    // Read locations from Trill sensor
+    for(unsigned int i = 0; i < activeTouches; i++) {
+      touchSize[i] = gTouchSensors[trillID]->touchSize(i);
+      touchVertLocation[i] = gTouchSensors[trillID]->touchLocation(i);
+      touchHoriLocation[i] = gTouchSensors[trillID]->touchHorizontalLocation(i);
+    }
+    // For all inactive touches, set location and size to 0
+    for(unsigned int i = activeTouches; i <  numTouches; i++) {
+      touchSize[i] = 0.0;
+      touchVertLocation[i] = 0.0;
+      touchHoriLocation[i] = 0.0;
+    }
+     
+    csnd::Vector<MYFLT> &out_size = outargs.vector_data<MYFLT>(1);
+    csnd::Vector<MYFLT> &out_v_location = outargs.vector_data<MYFLT>(2);
+    csnd::Vector<MYFLT> &out_h_location = outargs.vector_data<MYFLT>(3);
+
+    std::copy(touchSize, touchSize + numTouches, out_size.begin());
+    std::copy(touchVertLocation, touchVertLocation + numTouches, out_v_location.begin());
+    std::copy(touchHoriLocation, touchHoriLocation + numTouches, out_h_location.begin());
+    return OK;
+  }
+};
+
+void trillReadLoop(void*)
+{
+  while(!Bela_stopRequested())
+  {
+    for(unsigned int n = 0; n < gTouchSensors.size(); ++n)
+    {
+      Trill* t = gTouchSensors[n];
+      t->readI2C();
+    }
+    usleep(50000);
```

Comment:
This could be shorter, 10 to 15 millisecond would be ideal, but it should become smaller when many sensors are connected . Could you maybe expose this as a control parameter somehow?

Reply:
Hmm, it would be best to have that as a separate opcode then I think? Since you can have multiple instances of the trill opcodes, you dont want all of them to adjust this speed. For now, I will lower it